engineapi/snapshot.go

```diff
@@ -105,12 +105,14 @@ func (e *EngineBinary) SnapshotPurgeStatus(*longhorn.Engine) (map[string]*longho
 
 // SnapshotClone calls engine binary
 // TODO: Deprecated, replaced by gRPC proxy
-func (e *EngineBinary) SnapshotClone(engine *longhorn.Engine, snapshotName, fromControllerAddress string, fileSyncHTTPClientTimeout int64) error {
-	args := []string{"snapshot", "clone", "--snapshot-name", snapshotName, "--from-controller-address", fromControllerAddress}
+func (e *EngineBinary) SnapshotClone(engine *longhorn.Engine, snapshotName, fromEngineAddress, fromEngineName string,
+	fileSyncHTTPClientTimeout int64) error {
+	args := []string{"snapshot", "clone", "--snapshot-name", snapshotName, "--from-controller-address",
+		fromEngineAddress, "--from-controller-instance-name", fromEngineName}
```

Comment:
Looks this flag will break the old Longhorn engine version if we don't check engineCLIAPIVersion

Also, should we add the `--engine-instance-name`, `--replica-instance-name` CLI flags to other operation in `backups.go`, `engine.go`, `snapshot.go`,...

Reply:
I definitely need to either remove these args or add the version check.

Reply:
Discussed with Phan offline. Made local changes to add flags and CLIAPIVersion checks for this and similar code paths. I need some more time to test.

Reply:
Addressed this in my latest commit, but am finding it very difficult to test. It is not possible to launch an old enough instance-manager (v1.2.x) with a new longhorn-manager because the CLI commands we use to start the instance-manager have changed. The only way to do it is to actually install v1.2.x, create some volumes, and upgrade Longhorn all the way to master.

While working on this, I hit https://github.com/longhorn/longhorn-manager/pull/1897#issuecomment-1648668302.

Reply:
Addressed in https://github.com/longhorn/longhorn-manager/pull/1897/commits/fdb42152f48452d710ee0dcbd359db7b6c091d61.

Tested by:

1. Install Longhorn v1.2.x.
2. Deploy longhorn/examples/deployment.yaml and longhorn/examples/statefulset.yaml to get some volumes.
3. Upgrade stepwise through each version to master-head with the changes in this set of PRs.
4. Live engine upgrade the volumes to this set of PRs. This is the only way I know of to get the latest engine running in an old enough instance manager.

```
eweber@laptop:~/longhorn> kl get volume -oyaml | grep engineImage
    engineImage: webericjames/longhorn-engine:7dd52065
    engineImage: webericjames/longhorn-engine:7dd52065
    engineImage: webericjames/longhorn-engine:7dd52065

eweber@laptop:~/longhorn> kl get -oyaml lhim instance-manager-e-82604239
spec:
  image: longhornio/longhorn-instance-manager:v1_20220303_patch2
  nodeID: eweber-v124-worker-b8506f12-vwdng
  type: engine
status:
  apiMinVersion: 1
  apiVersion: 1
  currentState: running
  instances:
    pvc-c9d00a66-296f-4d88-a6b8-ab7c4006f646-e-6299b135:
      spec:
        backendStoreDriver: v1
        name: pvc-c9d00a66-296f-4d88-a6b8-ab7c4006f646-e-6299b135
      status:
        endpoint: ""
        errorMsg: ""
        listen: ""
        portEnd: 10001
        portStart: 10001
        resourceVersion: 0
        state: running
        type: engine
		
eweber@laptop:~/longhorn> kl get -oyaml lhim instance-manager-e-ffea61fd
spec:
  image: longhornio/longhorn-instance-manager:v1_20220303_patch2
  nodeID: eweber-v124-worker-b8506f12-9v4cr
  type: engine
status:
  apiMinVersion: 1
  apiVersion: 1
  currentState: running
  instances:
    pvc-9baa40fe-16b7-4823-bf50-550183783496-e-2de588d6:
      spec:
        backendStoreDriver: v1
        name: pvc-9baa40fe-16b7-4823-bf50-550183783496-e-2de588d6
      status:
        endpoint: ""
        errorMsg: ""
        listen: ""
        portEnd: 10001
        portStart: 10001
        resourceVersion: 0
        state: running
        type: engine
    pvc-876359cc-2569-4505-bf05-84675fb26548-e-b70657b2:
      spec:
        backendStoreDriver: v1
        name: pvc-876359cc-2569-4505-bf05-84675fb26548-e-b70657b2
      status:
        endpoint: ""
        errorMsg: ""
        listen: ""
        portEnd: 10002
        portStart: 10002
        resourceVersion: 0
        state: running
        type: engine
```

6. Verify that the live upgraded engine process are using the engine-identity-validation flags by execing into the instance-manager pods.

```
root@instance-manager-e-ffea61fd:/# ps -ef | cat
UID          PID    PPID  C STIME TTY          TIME CMD
root         180       8  0 15:52 ?        00:00:00 /engine-binaries/webericjames-longhorn-engine-7dd52065/longhorn --engine-instance-name pvc-876359cc-2569-4505-bf05-84675fb26548-e-b70657b2 controller pvc-876359cc-2569-4505-bf05-84675fb26548 --frontend tgt-blockdev --upgrade --replica tcp://10.42.66.11:10000 --replica tcp://10.42.137.157:10000 --replica tcp://10.42.186.20:10000 --size 1073741824 --current-size 1073741824 --engine-replica-timeout 8 --file-sync-http-client-timeout 30 --listen 0.0.0.0:10002
root         231       8  0 15:53 ?        00:00:00 /engine-binaries/webericjames-longhorn-engine-7dd52065/longhorn --engine-instance-name pvc-9baa40fe-16b7-4823-bf50-550183783496-e-2de588d6 controller pvc-9baa40fe-16b7-4823-bf50-550183783496 --frontend tgt-blockdev --upgrade --replica tcp://10.42.186.20:10010 --replica tcp://10.42.137.157:10010 --replica tcp://10.42.66.11:10010 --size 2147483648 --current-size 2147483648 --engine-replica-timeout 8 --file-sync-http-client-timeout 30 --listen 0.0.0.0:10001
```

7. Try to create a snapshot.

```
eweber@laptop:~/longhorn> kl get snapshot
NAME                                   VOLUME                                     CREATIONTIME           READYTOUSE   RESTORESIZE   SIZE       AGE
2d16a135-65a7-4748-961d-2c38b1f01efd   pvc-876359cc-2569-4505-bf05-84675fb26548   2023-07-25T15:57:19Z   true         1073741824    51003392   83s
```

8. Try to create a backup.

```
eweber@laptop:~/longhorn> kl get backup
NAME                      SNAPSHOTNAME                           SNAPSHOTSIZE   SNAPSHOTCREATEDAT      STATE       LASTSYNCEDAT
backup-91a91985cf50455d   2d16a135-65a7-4748-961d-2c38b1f01efd   62914560       2023-07-25T16:00:56Z   Completed   2023-07-25T16:01:01Z
```

9. Try to clone a volume.

```
eweber@laptop:~/longhorn> kl get volume pvc-c731ff07-cc04-4cad-8bc5-59a525b6b24b -oyaml
status:
  actualSize: 223137792
  cloneStatus:
    snapshot: 91680968-3282-5e6c-8771-1c8a3feacdd8
    sourceVolume: pvc-9baa40fe-16b7-4823-bf50-550183783496
    state: completed
```

10. Try to scale up a volume.

```
eweber@laptop:~/longhorn> kl get replica | grep pvc-876359cc-2569-4505-bf05-84675fb26548
pvc-876359cc-2569-4505-bf05-84675fb26548-r-6773d998   running   eweber-v124-worker-b8506f12-9v4cr   6887495f-17ba-4bff-bf41-fe20d2d1c777   instance-manager-1cd8b01fe2b1a8369f7d71d3e578279b   webericjames/longhorn-engine:9f790fdb   45m
pvc-876359cc-2569-4505-bf05-84675fb26548-r-c528b66a   running   eweber-v124-worker-b8506f12-2ngvh   39f9e765-7755-49a7-b37d-3b1e90c16b06   instance-manager-684405dba09a371ec21f1542a1ce4372   webericjames/longhorn-engine:9f790fdb   45m
eweber@laptop:~/longhorn> kl get replica | grep pvc-876359cc-2569-4505-bf05-84675fb26548
pvc-876359cc-2569-4505-bf05-84675fb26548-r-6773d998   running   eweber-v124-worker-b8506f12-9v4cr   6887495f-17ba-4bff-bf41-fe20d2d1c777   instance-manager-1cd8b01fe2b1a8369f7d71d3e578279b   webericjames/longhorn-engine:9f790fdb   45m
pvc-876359cc-2569-4505-bf05-84675fb26548-r-c528b66a   running   eweber-v124-worker-b8506f12-2ngvh   39f9e765-7755-49a7-b37d-3b1e90c16b06   instance-manager-684405dba09a371ec21f1542a1ce4372   webericjames/longhorn-engine:9f790fdb   45m
pvc-876359cc-2569-4505-bf05-84675fb26548-r-d01a3b56   running   eweber-v124-worker-b8506f12-vwdng   adb40a5b-a146-4823-b55a-da8a2fbcb7e9   instance-manager-37ee3914a4f3ebb7881e1fa1d67d5505   webericjames/longhorn-engine:9f790fdb   19s
eweber@laptop:~/longhorn> kl get volume | grep pvc-876359cc-2569-4505-bf05-84675fb26548
pvc-876359cc-2569-4505-bf05-84675fb26548   attached   healthy                  1073741824   eweber-v124-worker-b8506f12-9v4cr   5h27m
```